<a href="https://colab.research.google.com/github/DemydovaAnna/python_for_da_tasks/blob/main/Copy_of_Additional_hw_4_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Додаткове завдання з pandas
Вітаю! В цьому завданні будемо працювати з даними з опитування щодо зарплат ІТ спеціалістів на рексурсі DOU від грудня 2022 року. Впевнена, це буде цікаве дослідження, а заразом і практика pandas :)
Цей аналіз може бути корисним людині, яка обирає напрям роботи в ІТ і хоче зрозуміти тенденції. За якісного виконання цього аналізу, цей проєкт можна сміливо додавати у Ваше портфоліо!

1. Зчитайте дані з файлу `2022_dec_raw.csv`. Увага! Тут необхідно використати аргумент `pd.read_csv` `sep=';'`, бо розділювач в даних - крапка з комою.

In [48]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [49]:
import pandas as pd
df = pd.read_csv('drive/MyDrive/Dataloveacademy/Модуль 10. Бібліотека Pandas: ключові концепції та методи роботи з даними/2022_dec_raw.csv', sep=';')
df.head()

,Timestamp,Ви працюєте зараз в ІТ?,Де ви зараз живете?,В якій області ви зараз живете?,"Зарплата у $$$ за місяць, лише ставка після сплати податків",Ваша стать,Ваш вік,Яка у вас освіта?,Чи є у вас науковий ступінь?,Знання англійської мови,...,Ваша посада4,Ваша посада5,Ваша посада6,Ваша посада7,Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?,Основна мова програмування,Вкажіть вашу основну спеціалізацію,В якій сфері працюєте?,Тип компанії,Загальний стаж роботи за спеціальністю
0,11.30.2022 10:26:37,"Так, працюю part-time",В Україні,Івано-Франківськ чи область,4000,Чоловік,29,"Вища (бакалавр, спеціаліст, магістр)",Немає,Intermediate,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Blockchain,Стартап,8 років
1,11.30.2022 10:26:39,"Так, працюю full-time","За кордоном (переїхали через війну, але планую...",NaN,NaN,Чоловік,36,"Вища (бакалавр, спеціаліст, магістр)",Немає,Advanced,...,NaN,NaN,NaN,NaN,NaN,Ruby,Full Stack розробка,Public services / Government,Продуктова,3 роки
2,11.30.2022 10:26:58,"Так, працюю full-time",В Україні,Київ чи область,2500,Чоловік,34,"Вища (бакалавр, спеціаліст, магістр)",Немає,Elementary,...,NaN,NaN,NaN,NaN,NaN,C# / .NET,Full Stack розробка,"Adtech / Advertising, Security, Telecom",Продуктова,4 роки
3,11.30.2022 10:27:31,"Так, працюю full-time",В Україні,Львів чи область,500,Жінка,23,"Вища (бакалавр, спеціаліст, магістр)",Немає,Pre-Intermediate,...,NaN,NaN,NaN,NaN,"Ні, не використовую",NaN,NaN,Telecom,Продуктова,Пів року
4,11.30.2022 10:27:34,"Ні, тимчасово не працюю",В Україні,Житомир чи область,2800,Жінка,33,"Вища (бакалавр, спеціаліст, магістр)",Немає,Upper-Intermediate,...,NaN,NaN,NaN,NaN,"Ні, не використовую",NaN,NaN,Automotive,Продуктова,6 років


2. Які колонки мають дані?

In [50]:
df.columns

Index(['Timestamp', 'Ви працюєте зараз в ІТ?', 'Де ви зараз живете? ',
       'В якій області ви зараз живете?',
       'Зарплата у $$$ за місяць, лише ставка після сплати податків',
       'Ваша стать', 'Ваш вік', 'Яка у вас освіта?',
       'Чи є у вас науковий ступінь? ', 'Знання англійської мови',
       'Оберіть вашу посаду', 'Ваш тайтл', 'Ваша спеціалізація',
       'Вкажіть вашу спеціалізацію', 'Оберіть вашу спеціалізацію',
       'Ваша посада', 'Оберіть вашу посаду2', 'Ваша посада3', 'Ваша посада4',
       'Ваша посада5', 'Ваша посада6', 'Ваша посада7',
       'Чи використовуєте ви у своїй роботі мови програмування (одну чи декілька)?',
       'Основна мова програмування', 'Вкажіть вашу основну спеціалізацію',
       'В якій сфері працюєте?', 'Тип компанії',
       'Загальний стаж роботи за спеціальністю'],
      dtype='object')

3. Скільки рядків та колонок мають дані? Скільки пропущених значень має кожна з колонок?

In [51]:
print(f"Кількість рядків: {df.shape[0]}")
print(f"Кількість колонок: {df.shape[1]}\n")
df.isna().sum()

Кількість рядків: 13315
Кількість колонок: 28



,0
Timestamp,0
Ви працюєте зараз в ІТ?,0
Де ви зараз живете?,0
В якій області ви зараз живете?,1777
"Зарплата у $$$ за місяць, лише ставка після сплати податків",27
Ваша стать,0
Ваш вік,0
Яка у вас освіта?,0
Чи є у вас науковий ступінь?,111
Знання англійської мови,0


4. Які колонки мають всі значення заповненими (не мають дожного None значення)? Виведіть програмно список цих колонок.

In [52]:
not_null_cols = df.columns[df.notna().all()].tolist()
print("Колонки без жодного пропускa:")
print(not_null_cols)

Колонки без жодного пропускa:
['Timestamp', 'Ви працюєте зараз в ІТ?', 'Де ви зараз живете? ', 'Ваша стать', 'Ваш вік', 'Яка у вас освіта?', 'Знання англійської мови', 'Оберіть вашу посаду', 'Ваш тайтл', 'В якій сфері працюєте?', 'Тип компанії', 'Загальний стаж роботи за спеціальністю']


5. Який тип мають колонки в наборі даних? Які з колонок типу object логічно на ваш погляд перетворити на числові колонки?

In [53]:
df.dtypes

,0
Timestamp,object
Ви працюєте зараз в ІТ?,object
Де ви зараз живете?,object
В якій області ви зараз живете?,object
"Зарплата у $$$ за місяць, лише ставка після сплати податків",object
Ваша стать,object
Ваш вік,object
Яка у вас освіта?,object
Чи є у вас науковий ступінь?,object
Знання англійської мови,object


Більшість колонок мають тип ⁠object⁠ (текстовий).
Колонки типу ⁠object⁠, які логічно перетворити на числові:

1. Ваш вік та Зарплата у $$$ за місяць, лише ставка після сплати податків це числові дані, які
через текстовий формат зчиталися
як object . Іх потрібно перетворити
на float .
2. Timestamp - це текстовий запис дати й
часу, який логічно перетворити на тип datetime для зручної роботи з
часовими рядами.


6. Трансформуйте колонки 'Ваш вік', 'Зарплата у $$$ за місяць, лише ставка після сплати податків' у тип даних float. УВАГА! Перед трансформацією подивіться, які дані містять ці колонки. Аби успішно трансформувати тип даних, треба спочатку привести рядки до того виду, який буде підходити, аби потім рядок зробити типом float.

In [54]:
df['Ваш вік'] = df['Ваш вік'].astype(str).str.replace(',', '.').astype(float)
df['Зарплата у $$$ за місяць, лише ставка після сплати податків'] = df['Зарплата у $$$ за місяць, лише ставка після сплати податків'].astype(str).str.replace(',', '.').astype(float)

df[['Ваш вік', 'Зарплата у $$$ за місяць, лише ставка після сплати податків']].dtypes

,0
Ваш вік,float64
"Зарплата у $$$ за місяць, лише ставка після сплати податків",float64


7. Це завдання з зірочкою! Нам треба транфсформувати колонку Timestamp у тип даних `datetime64[ns]`. Це не так просто як здається на першйий погляд :) Але це реальна ситуація, яка може трапитись при зборі даних в роботі аналітиком.
Я запропоную один зі способів виконати завдання, але ви можете написати по-своєму. Головне - отримати колонку - потрібного типу.
Пропоную наступний алгоритм дій:
    1. Подивитись, які дані містяться в цій колонці. Виведіть перші і останні 5 записів. Чи однаковий формат цих даних? (спойлер: ні, бо інакше все було б просто :))
    2. Нам треба враховуючи різний формат даних перетворити рядок у колонці `Timestamp` на тип `datetime64[ns]`. Як це зробити? Для перетворення рядка на `datetime64` пропоную скористатись `datetime.datetime.strptime`.
    Для цього треба імпортувати бібліотеку `datetime`. Я пропоную написати функцію для обробки кожного окремого рядка, яка враховує довжину рядка і на основі цього застосовує різний формат даних дати.
    Як прописувати формат даних можна переглянути [тут](https://www.w3schools.com/python/python_datetime.asp).
    3. Коли функція для обробки кожного окремого рядка готова, застосуйте її для всіх даних з допомогою методу `apply`.

In [55]:
print("Перші 5 записів:")
print(df['Timestamp'].head(5))

print("\nОстанні 5 записів:")
print(df['Timestamp'].tail(5))

Перші 5 записів:
0    11.30.2022 10:26:37
1    11.30.2022 10:26:39
2    11.30.2022 10:26:58
3    11.30.2022 10:27:31
4    11.30.2022 10:27:34
Name: Timestamp, dtype: object

Останні 5 записів:
13310    01.09.2023 11:01
13311    01.09.2023 11:14
13312    01.09.2023 11:19
13313    01.09.2023 11:28
13314    01.09.2023 11:38
Name: Timestamp, dtype: object


In [56]:
import re
patterns = df['Timestamp'].astype(str).apply(lambda x: re.sub(r'\d', 'd', x)).unique()
patterns

array(['dd.dd.dddd dd:dd:dd', 'dd.dd.dddd dd:dd', 'dd.dd.dddd d:dd:dd'],
      dtype=object)

In [57]:
import pandas as pd
def parse_timestamp_by_len(val):
  val = str(val).strip()
  if len(val) == 19:
    return datetime.strptime(val, '%m.%d.%Y %H:%M:%S')
  elif len(val) == 16:
    return datetime.strptime(val, '%m.%d.%Y %H:%M')
  elif len(val) == 18:
    return datetime.strptime(val, '%m.%d.%Y %H:%M:%S')
  return None
df['Timestamp'] = df['Timestamp'].apply(parse_timestamp_by_len)
df['Timestamp'].dtype

dtype('<M8[ns]')

In [58]:
import re
patterns = df['Timestamp'].astype(str).apply(lambda x: re.sub(r'\d', 'd', x)).unique()
patterns

array(['dddd-dd-dd dd:dd:dd'], dtype=object)

8. Для колонок типу даних object (які такими лишились після всіх наших трансформацій) виведіть кількість унікальних значень.

In [59]:
object_cols = df.select_dtypes(include=['object']).columns
for col in object_cols:
  print(f"Колонка '{col}': {df[col].nunique()} унікальних значень")

Колонка 'Ви працюєте зараз в ІТ?': 5 унікальних значень
Колонка 'Де ви зараз живете? ': 2 унікальних значень
Колонка 'В якій області ви зараз живете?': 23 унікальних значень
Колонка 'Ваша стать': 2 унікальних значень
Колонка 'Яка у вас освіта?': 5 унікальних значень
Колонка 'Чи є у вас науковий ступінь? ': 3 унікальних значень
Колонка 'Знання англійської мови': 5 унікальних значень
Колонка 'Оберіть вашу посаду': 22 унікальних значень
Колонка 'Ваш тайтл': 10 унікальних значень
Колонка 'Ваша спеціалізація': 8 унікальних значень
Колонка 'Вкажіть вашу спеціалізацію': 4 унікальних значень
Колонка 'Оберіть вашу спеціалізацію': 21 унікальних значень
Колонка 'Ваша посада': 15 унікальних значень
Колонка 'Оберіть вашу посаду2': 7 унікальних значень
Колонка 'Ваша посада3': 7 унікальних значень
Колонка 'Ваша посада4': 5 унікальних значень
Колонка 'Ваша посада5': 6 унікальних значень
Колонка 'Ваша посада6': 2 унікальних значень
Колонка 'Ваша посада7': 4 унікальних значень
Колонка 'Чи використовуєте

9. Для колонок типу даних object виведіть топ 10 найбільш частих значень з кількістю рядків, в яких це значення зустрічається. Зробіть це для кожної колонки окремо. Наприклад, в циклі.  
Що цікавого бачите? Які висновки можна зробити? Що особисто Вас вразило?

In [60]:
object_cols = df.select_dtypes(include=['object']).columns
for col in object_cols:
  print(f"=== Колонка: {col} ===")
  print(df[col].value_counts().head(10))
  print("\n" + "="*40 + "\n")

=== Колонка: Ви працюєте зараз в ІТ? ===
Ви працюєте зараз в ІТ?
Так, працюю full-time      12643
Ні, тимчасово не працюю      380
Так, працюю part-time        152
Так, я фрилансер             115
Інше                          25
Name: count, dtype: int64


=== Колонка: Де ви зараз живете?  ===
Де ви зараз живете? 
В Україні                                                      11587
За кордоном (переїхали через війну, але планую повернутися)     1728
Name: count, dtype: int64


=== Колонка: В якій області ви зараз живете? ===
В якій області ви зараз живете?
Київ чи область                4411
Львів чи область               2353
Дніпро чи область               751
Івано-Франківськ чи область     478
Вінниця чи область              447
Одеса чи область                443
Харків чи область               355
Полтава чи область              252
Черкаси чи область              232
Хмельницький чи область         230
Name: count, dtype: int64


=== Колонка: Ваша стать ===
Ваша стать
Чоловік  

Що цікавого:
1. Гендерний дисбаланс та освіта:
 У вибірці суттєво переважають чоловіки (10,221 проти 3,094 жінок), що демонструє стійкий гендерний дисбаланс в українському IT.
 Абсолютна більшість респондентів має вищу освіту (10,645 осіб), а науковий ступінь (PhD / кандидат наук) є рідкістю (лише 216 осіб).
2. Географія та релокація:
 Понад 1,728 айтівців вказали, що виїхали за кордон через війну.
 Серед тих, хто залишається в Україні, ключовими хабами є Київ та область (4,411) і Львів та область (2,353), далі йдуть Дніпро, Івано-Франківськ та Вінниця.
3. Посади та технології:
 Найчисельніша категорія посадових осіб — Software Engineer (6,773), на другому місці — QA Engineer (2,364).
 Серед мов програмування лідирує JavaScript (1,695), за ним йдуть Java (1,241), Python (1,190), C# / .NET (1,179) та TypeScript (1,179).
 Найбільше фахівців зайняті в Back-end (2,847) та Full Stack (1,628) розробці.
4. Структура бізнесу:
 Сфера українського IT тримається на Аутсорсингових (5,593) та Продуктових (4,940) компаніях.
 Домінуючими сферами є E-commerce (1,243) та Fintech / Banking (942).


Основні висновки з даних:

 Географія та міграція: Війна суттєво змінила IT-карту — понад 1 700 айтівців (понад 13% вибірки) виїхали за кордон, а всередині країни Львів посів друге місце після Києва, обігнавши Харків та Дніпро.
 Домінування розробки та QA: Понад 68% респондентів — це Software Engineers та QA Engineers.

Що особисто мене вражає в цих даних:

 Низька частка наукових ступенів при вищій освіті: Понад 80% респондентів мають вищу освіту (бакалавр/магістр), але лише 1.6% (216 осіб) мають PhD або є кандидатами наук. Це демонструє, що в IT цінність практичних навичок значно вища за академічну кар'єру.

 Низька частка фрілансу: фрілансерів у вибірці лише 121 особа проти ~10 500 працівників в Аутсорсі та Продукті.

10. Виведіть медіанну (медіана - це значення, менше якого є рівно 50% вибірки і більше якого є теж рівно 50% вибірки)
ЗП (Зарплата у $$$) для кожного рівня знання англійської. Які висновки можна зробити?

In [63]:
median_salary = df.groupby('Знання англійської мови')['Зарплата у $$$ за місяць, лише ставка після сплати податків'].median().reset_index()
median_salary.columns = ['Рівень англійської', 'Медіанна ЗП ($)']
print(median_salary)

   Рівень англійської  Медіанна ЗП ($)
0            Advanced           3500.0
1          Elementary           1155.0
2        Intermediate           2125.0
3    Pre-Intermediate           1404.0
4  Upper-Intermediate           3300.0


Спостерігається чітка позитивна кореляція: чим вищий рівень англійської, тим вища медіанна зарплата фахівця.

 Фахівці з рівнями Advanced та Upper-Intermediate заробляють суттєво більше за тих, хто має Elementary чи Pre-Intermediate, оскільки високий рівень мови дозволяє працювати безпосередньо з іноземними замовниками (США, ЄС), де ставки за годину значно вищі, а також посади високих рівнів вимагають постійної комунікації англійською, тому високий рівень мови є обов'язковою умовою для високооплачуваних позицій.

11. Є така штука як сортування значень категоріальної змінної. Нам це зараз стане в нагоді, адже рівні англійської можна посортувати.
Давайте відсотруємо рівні англ. і тайтли. Просто запустіть код нижче і проаналізуйте, що він робить.

In [64]:
df['Знання англійської мови'] = pd.Categorical(df['Знання англійської мови'],
                      categories=['Elementary','Pre-Intermediate','Intermediate','Upper-Intermediate','Advanced'],
                      ordered=True)

df['Ваш тайтл'] = pd.Categorical(df['Ваш тайтл'],
                      categories=['Intern/Trainee', 'Junior', 'Немає тайтлу', 'Middle', 'Senior', 'Tech Lead', 'Team Lead', 'Architect', 'Head', 'Manager'],
                      ordered=True)

12. Зробіть зводну таблицю, де стовпці - це 'Ваш тайтл', рядки - 'Знання англійської мови', а значення - медіанна ЗП. Які висновки можна зробити?
Тут ще для зручності можна зробити градієнтне забарвлення і відформатувати значення для легшого перегляду.

In [65]:
pivot_table = df.pivot_table(
    index='Знання англійської мови',
    columns='Ваш тайтл',
    values='Зарплата у $$$ за місяць, лише ставка після сплати податків',
    aggfunc='median',
    observed=False
)
pivot_table.style.background_gradient(cmap='YlGn').format('{:.0f}', na_rep='-')

Ваш тайтл,Intern/Trainee,Junior,Немає тайтлу,Middle,Senior,Tech Lead,Team Lead,Architect,Head,Manager
Знання англійської мови,,,,,,,,,,
Elementary,500,600,550,1500,2500,5000,2100,-,1500,1120
Pre-Intermediate,400,800,934,1825,3500,4000,2850,2900,2400,1225
Intermediate,400,900,1300,2200,4100,5000,3800,5450,3500,2100
Upper-Intermediate,485,1000,1600,2500,4750,5650,4655,6400,4000,3600
Advanced,500,1000,2075,2500,4750,6000,5000,7000,5000,4000


Тайтл залишається головним фактором росту ЗП. Перехід від Junior до Senior дає кратне зростання доходу незалежно від мови.

Англійська мова працює як множник: на кожному тайтлі (особливо Middle та Senior) фахівці з Advanced / Upper-Intermediate мають помітно вищу медіанну ЗП, ніж їхні колеги з Elementary.

13. Виведіть результат методу describe() по колонцы розміру ЗП для топ 5 за частотою зустрічання в даних областей (В якій області ви зараз живете?). Зробіть висновки, де вигідніше жити з точки зору ЗП і чи суттєва різниця на Ваш погляд для прийняття рішення про переїзд?

In [69]:
top5_regions = df['В якій області ви зараз живете?'].value_counts().head(5).index
df_top5_regions = df[df['В якій області ви зараз живете?'].isin(top5_regions)]
df_top5_regions.groupby('В якій області ви зараз живете?')['Зарплата у $$$ за місяць, лише ставка після сплати податків'].describe().round(2).sort_values(by='count', ascending=False)

,count,mean,std,min,25%,50%,75%,max
В якій області ви зараз живете?,,,,,,,,
Київ чи область,4411.0,3141.81,2039.86,80.0,1500.0,2800.0,4500.0,16000.0
Львів чи область,2353.0,3121.44,2102.72,150.0,1375.0,2700.0,4500.0,14000.0
Дніпро чи область,751.0,2582.58,1864.71,100.0,1025.0,2000.0,3600.0,12000.0
Івано-Франківськ чи область,478.0,2705.26,1905.67,200.0,1200.0,2372.5,3775.0,12436.0
Вінниця чи область,447.0,2888.31,1902.67,100.0,1300.0,2612.0,4100.0,11675.0


Аналіз рівнів ЗП по регіонах:

Київ та Львів є беззаперечними лідерами як за чисельністю фахівців, так і за рівнями доходів. Медіанна ЗП становить *\$2800* у Києві та *\$2700* у Львові, а середня (⁠mean⁠) перетинає позначку в $3100.

Дніпро, Івано-Франківськ та Вінниця демонструють нижчі медіанні показники: від *\$2000* (Дніпро) до $2612.50 (Вінниця).

Де вигідніше жити з точки зору ЗП:

З суто фінансової точки зору найвищий стелю доходів пропонують Київ та Львів (максимальні ЗП сягають *\$14 000 - $16 000*).

Переїзд не є обов'язковим, оскільки IT-ринок в Україні часто працює у віддаленому форматі. Дохід залежить від кваліфікації та компанії замовника, а не від фізичного міста.

Вищий дохід у Києві чи Львові нівелюється значно вищою вартістю оренди житла та життя порівняно з регіонами.

14. Як зростає зарплата з роками? Для цього нам треба відформатувати колонку "Загальний стаж роботи за спеціальністю" аналогічно до того, як ми це робили для колонки 'Знання англійської мови' і потім вивести медіанні значення для кількості років стажу.

In [71]:
df['Загальний стаж роботи за спеціальністю'].unique()

array(['8 років', '3 роки', '4 роки', 'Пів року', '6 років', '1,5 року',
       '1 рік', '2 роки', '5 років', '10 років', 'Менше як 3 місяці',
       '11 років', '7 років', '15 і більше років', '3 місяці', '12 років',
       '9 років', '14 років', '13 років'], dtype=object)

In [72]:
exp_categories = [
    'Менше як 3 місяці',
    '3 місяці',
    'Пів року',
    '1 рік',
    '1,5 року',
    '2 роки',
    '3 роки',
    '4 роки',
    '5 років',
    '6 років',
    '7 років',
    '8 років',
    '9 років',
    '10 років',
    '11 років',
    '12 років',
    '13 років',
    '14 років',
    '15 і більше років'
]
df['Загальний стаж роботи за спеціальністю'] = pd.Categorical(
    df['Загальний стаж роботи за спеціальністю'],
    categories=exp_categories,
    ordered=True )
experience_salary = df.groupby('Загальний стаж роботи за спеціальністю', observed=False)['Зарплата у $$$ за місяць, лише ставка після сплати податків'].median().reset_index()
experience_salary.columns = ['Стаж роботи', 'Медіанна ЗП ($)']
experience_salary

,Стаж роботи,Медіанна ЗП ($)
0,Менше як 3 місяці,800.0
1,3 місяці,525.0
2,Пів року,700.0
3,1 рік,950.0
4,"1,5 року",1200.0
5,2 роки,1700.0
6,3 роки,2450.0
7,4 роки,3000.0
8,5 років,3500.0
9,6 років,4000.0


Динаміка зростання доходів від стажу:

 Найбільш динамічне зростання спостерігається в перші 5 років роботи. Медіанна ЗП зростає від *\$950* (1 рік) до *\$3500* (5 років), що відображає швидку зміну кваліфікаційних рівнів від Junior до Senior.

У діапазоні від 6 до 10 років стажу зростання сповільнюється — медіана коливається в межах *\$4000* - *\$4700*.

Після 10 років стажу медіанна ЗП стабілізується на рівні *\$4500 - $5000*. Незначні коливання в окремих роках (наприклад, 13 чи 15+ років) зумовлені меншою кількістю респондентів у цих вибірках або переходом фахівців у суміжні управлінські ролі, де структура доходу може відрізнятися.

Досвід є ключовим драйвером доходу на початкових та середніх етапах кар'єри, тоді як після 5–7 років досвіду вирішальну роль починають відігравати тайтл (Senior / Lead / Architect) та рівень володіння англійською мовою.

Далі тут можна зробити багато аналогічних агрегацій, проаналізувати, в який час люди заповнювали форму, хто більше отримує, Business Analyst, чи Data Analyst. Я вас заохочувати не стримувати свій інтерес і покапати ці дані далі :)